### PR TITLE
Test delegation store

### PR DESCRIPTION
**Pull request by expede**
Wrote some tests for the (now fixed?) delegation store. It still has some commented-out work for the predicate language, but that is queued up for after we get the basics working for the server.

# Changes

* Fix logic in `get_chain`
* Add some delegation store tests
* Add a `store.insert(delegation)` method as a convenience to not need to take the CID manually
* Test the underlying `jq`-style `Selector` logic that was causing issues at the higher levels
* Break up `delegation::store` into `delegation::store::memory` and `::traits`, just like the other stores

### CODE REVIEW

File: src/delegation/agent.rs [hunk 1]
```diff
@@ -20,29 +20,34 @@ use web_time::SystemTime;
...
-    Enc: Codec + Into<u64> + TryFrom<u64> = varsig::encoding::Preset,
-> {
+    S: Store<DID, V, C>,
+    DID: Did + Clone = did::preset::Verifier,
+    V: varsig::Header<C> + Clone = varsig::header::Preset,
+    C: Codec + Into<u64> + TryFrom<u64> = varsig::encoding::Preset,
+> where
+    Delegation<DID, V, C>: Encode<C>,
+    Payload<DID>: TryFrom<Named<Ipld>>,
+    Named<Ipld>: From<Payload<DID>>,
```
Comment (expede) [inline src/delegation/agent.rs]: I'm going to look at removing some or all of these, but this is required to have a `.insert` method that doesn't ask for the CID manually (because it needs to be able to encode the CID)

File: src/delegation/agent.rs [hunk 2]
```diff
@@ -20,29 +20,33 @@ use web_time::SystemTime;
...
-> {
+    S: Store<DID, V, C>,
+    DID: Did + Clone = did::preset::Verifier,
+    V: varsig::Header<C> + Clone = varsig::header::Preset,
+    C: Codec + Into<u64> + TryFrom<u64> = varsig::encoding::Preset,
+> where
+    Ipld: Encode<C>,
+    Payload<DID>: TryFrom<Named<Ipld>>,
+    Named<Ipld>: From<Payload<DID>>,
+{
```
Comment (expede) [inline src/delegation/agent.rs:31]: The change in the above file using `to_ipld_envelope` was way more subtle than it should have been. basically previously the payload types individually implemented `libipld::Encode<C>`, but that was (evidently) confusing. It took longer than I'd like to admit to find the bug it caused.

File: src/delegation/policy/selector/select.rs [hunk 3]
```diff
@@ -46,9 +46,9 @@ impl<T> Select<T> {
 
 impl<T: Selectable> Select<T> {
     pub fn get(self, ctx: &Ipld) -> Result<T, SelectorError> {
```
Comment (expede) [inline src/delegation/policy/selector/select.rs:48]: The `Try` behaviour in this method could probably be more elegant, but I really don't want to have to write more custom `Try` (AKA `.?`) parsing 😅

File: src/delegation/store/memory.rs [hunk 4]
```diff
@@ -160,7 +160,7 @@ impl<
 where
     Named<Ipld>: From<delegation::Payload<DID>>,
     delegation::Payload<DID>: TryFrom<Named<Ipld>>,
-    Delegation<DID, V, Enc>: Encode<Enc>,
+    Ipld: Encode<Enc>,
```
Comment (expede) [inline src/delegation/store/memory.rs:163]: Noted in an Issue comment above

File: src/delegation/store/memory.rs [hunk 5]
```diff
@@ -173,7 +173,7 @@ where
         // FIXME
     }
 
-    fn insert(
+    fn insert_keyed(
```
Comment (expede) [inline src/delegation/store/memory.rs:176]: This lets us automagically calculate the CID on `insert`, but let the trait implementer work with the lower level interface rather than forcing them to figure out the CID stuff themselves.

File: src/delegation/store/memory.rs [hunk 6]
```diff
@@ -258,17 +257,19 @@ where
                             continue;
                         }
 
-                        for target_pred in policy.iter() {
-                            for delegate_pred in delegation.payload.policy.iter() {
-                                let comparison =
-                                    target_pred.harmonize(delegate_pred, vec![], vec![]);
+                        // FIXME
```
Comment (expede) [inline src/delegation/store/memory.rs:260]: Sitting in a commented-out `FIXME` until we write the predicate subsumption logic (which I'll switch back to after fixing a couple other things that I'm unblocking)

File: src/delegation/store/memory.rs [hunk 7]
```diff
@@ -281,34 +282,583 @@ where
...
+                        (deleg_1.cid()?, Arc::new(deleg_1)),
+                    ]
+                    .into()
+                )
+            );
+            Ok(())
+        }
+
+        #[test_log::test]
+        fn test_looking_for_narrower_command() -> TestResult {
```
Comment (expede) [inline src/delegation/store/memory.rs:575]: This is a copy of the original integration test, but without the `Agent` bits. i.e. testing a smaller surface.

File: src/invocation/store.rs [hunk 8]
```diff
@@ -1,134 +1,7 @@
...
-use super::Invocation;
-use crate::ability;
-use crate::{crypto::varsig, did::Did};
-use libipld_core::{cid::Cid, codec::Codec};
-use std::sync::{Arc, RwLock, RwLockReadGuard, RwLockWriteGuard};
-use std::{collections::BTreeMap, convert::Infallible};
+mod memory;
+mod traits;
 
-pub trait Store<T, DID: Did, V: varsig::Header<C>, C: Codec + Into<u64> + TryFrom<u64>> {
```
Comment (expede) [inline src/invocation/store.rs:10]: Just made the modules layout the same as all of the other stores. This probably should be in its own PR, but it was driving me up the wall while working 😅

File: src/delegation/store/traits.rs [hunk 9]
```diff
@@ -1,25 +1,39 @@
...
 use nonempty::NonEmpty;
 use std::{fmt::Debug, sync::Arc};
 use web_time::SystemTime;
 
-pub trait Store<DID: Did, V: varsig::Header<Enc>, Enc: Codec + TryFrom<u64> + Into<u64>> {
+pub trait Store<DID: Did + Clone, V: varsig::Header<C> + Clone, C: Codec + TryFrom<u64> + Into<u64>>
+where
+    Ipld: Encode<C>,
+    Payload<DID>: TryFrom<Named<Ipld>>,
+    Named<Ipld>: From<Payload<DID>>,
```
Comment (matheus23) [inline src/delegation/store/traits.rs:20]: This combination of constraints comes up repeatedly a bit.
Perhaps worth keeping in the back of our minds whether we should turn this into a trait that combines all these constraints? Personally I'd also appreciate getting away from `From` and `TryFrom` for parsing. IMO they can make code & compiler errors less readable.
Not in this PR though! Let's just keep it in mind.

File: src/invocation/store.rs (hunk 8 above)
Comment (matheus23) [reply src/invocation/store.rs:10]: Haha, same. I'd be working in `delegation/store/memory.rs` and search forever for the invocation store (because I would think it'd be in `invocation/store/memory.rs`, so I'd fuzzy-search "invocationmemory", but... :sweat_smile:)

**Review by matheus23 — APPROVED** [review]
